Bookmarks

cguglielmo opened the pull request.

bschwarzent committed:
Bookmarks: add bookmark support in JS

- Add BookmarkSupport to create and activate bookmarks in JS.
- Page: add support for page params in JS

389569

bschwarzent committed:
BookmarkSupport: export search data

389569

mvilliger committed:
Add PageResolver

389577

mvilliger committed:
Uuid for Scout JS objects

389577

cguglielmo committed:
Uuid for Scout JS objects: rename BookmarkAdapter

- A Widget should not have a BookmarkAdapter
- Java also has a BookmarkAdapter but it serves a different purpose
-> Rename to ObjectUuidBuilder

Also improve fallback logic of uuid generation: objectType may be
specified explicitly in Model-files. In that case we need to resole
the string based objectType first.

389577

cguglielmo committed:
DataObjects: add support for do type resolvers

Sometimes it is desired to use a specific BaseDoEntity class if the
data object _type cannot be resolved rather than the generic
BaseDoEntity. In that case a custom do type resolver can be added.

389577

cguglielmo committed:
Bookmarks: improve page param handling

- PageParamDo is now a class
- A PageParamDo instance is created automatically if _type ends
  with PageParam -> code does not need to handle pojo params.
- Some contributions are now removed when comparing params, same
  logic as in Java.
- Add BookmarkAdapter to make bookmark specific logic like compare logic
  extensible
- Add StringValueDo. ValueDo needs to stay an interface.

389569

bschwarzent committed:
Bookmarks: compute row identifier for js table pages only

For Classic table pages, the bookmarkIdentifier is generated on
the UI server and sent as a static property.

For JS table pages, the bookmark identifier is computed in JavaScript
when needed.

For normal tables (e.g. in a table field), no identifier has to be
generated.

389569

bschwarzent committed:
Bookmarks: improve method arguments in BookmarkSupport

389569

cguglielmo committed:
Tree: add page detail menu contributor

A page detail menu contributor can adapt the menus of a
Page.detailTable or Page.detailForm. This makes it possible to add
menus to a page without subclassing.

The existing logic to inherit the menus from a parent table page
has been extracted to a contributor.

Also:
- Add ObjectOrType to simplify some type definitions
- Simplify MenuSpecHelper by removing iconId

389569

cguglielmo committed:
Bookmarks: ensure classId and uuid match for JsPage

If a JsPage gets a uuid and already has a classId, the uuid needs to be
the same.

Use case:
- Add a bookmark for a js page that has a classId.
- Object ObjectUuidProvider returns classId because it has one.
- Open bookmark in bookmark table page -> Java tries to resolve the
  JS object type for the given PageParam containing the classId
  using JsPageUiHandler which uses PageResolver.ts
- PageResolver searches all JS subclasses of Page and returns the one
  matching the given classId. But: no page has that classId because
  they are not created by Java and no uuid matches.

Possible solutions:
- When creating the PageIdDummyPageParamDo with Java, it needs to use
  the uuid from JS using a UiCallback -> seems to be too much effort
- Uuid and classId should match -> seems reasonable
- ObjectUuidProvider needs to return uuid first, maybe only for JsPages
  -> May work, but maybe consistent ids may help for other cases as well

389577

cguglielmo committed:
JsonOutline: do not send detail table for invisible root node

If the invisible root node is set when the json outline already exists
(e.g. by using IDesktop.PROP_READY event), a page changed event is
fired and handled by JsonOutline. The JsonOutline does not send
the page change event to the UI because optNodeId returns null for
the invisible root node, but it attaches the detail table and form.
If a detail table is sent the UI may throw errors because elements
may refer to that table (e.g. ChartTableControl created by its
HelpBubble) but no page created it (because there is no invisible
root node in the UI).

cguglielmo committed:
JsonContextMenu: allow for custom structure changed handlers

bschwarzent committed:
UiCallbacks: add support for context elements (and general cleanup)

- Add possibility to send a set of "context elements" along with the
  UI callback and/or its response. The concept is very similar to the
  hybrid actions, except that the call is initiated by the server. The
  same data structures are re-used. Json/model conversion was extracted
  to helpers (JsonHybridHelper, hybridUtil).
- Clearly distinguish the in-memory result object of a UI callback
  handler from the actual data object that is sent back to the server.
- Use input/param objects to reduce the number of arguments.
- Code cleanup: use consistent naming and order for classes, methods,
  members and parameters.

389569

bschwarzent committed:
Bookmarks: allow activating bookmark from both the UI and the server

Bookmark activated from UI (JS):
1. Execute "ActivateBookmark" hybrid action.
2. Hybrid action tries to resolve as much of the bookmark as it can in
   Java. If no bookmark pages remain, we are done.
3. Otherwise, the remaining bookmark pages are sent back to the UI
   via the "OpenBookmarkInOutline" ui callback handler, along with
   parent Java page as context element.
4. The callback handler starts at the specified location and applies
   the remaining bookmark entries in JS.
5. The UI notifies the server that the ui callback is complete.
6. The server notifies the UI that the hybrid action is complete.

Bookmark activated from server (Java):
Exactly the same, except that no hybrid action is involved, i.e. steps
1 and 6 are not necessary.

389569

bschwarzent committed:
Bookmarks: add support for applying a bookmark to JS pages

389569

bschwarzent committed:
Bookmarks: add support for updating a JS page bookmark definition

389569

bschwarzent committed:
Bookmarks: consolidate applyBookmarkToPage

Page state of bookmarked pages should be applied at a single location
in code (currently BookmarkSupport#applyBookmarkToPage).

389569

bschwarzent committed:
Bookmarks: use async/await in BookmarkSupport

- Add support for outline-only bookmarks
- Allow resetting search form to default when opening bookmark in
  outline
- Add Table#expandParentRows to expand all parent rows

389569

bschwarzent committed:
Bookmarks: add BookmarkSupportSpec

389569

bschwarzent committed:
Bookmarks: fix error when bookmarked page has no search form

389569

cguglielmo committed:
Bookmarks: improve pageParam handling

- PageParam for JsPage
  A JsPage can now resolve its object type based on its pageParam
- get/set pageParam removed
  This makes it possible to declare the type of the param without the
  need to override get AND set. Also having a _setPageParam is a more
  common pattern in Scout than using a setter.
- Made pageParam mandatory in initModel of the page if the page
  declares an explicit pageParam and does not use the default
  PageParamDo
- Deprecated the _pageParam method
  It has no real benefits and the name conflicts with the new pageParams
- Replaced Page.getOutline() with Page.outline

389569

cguglielmo committed:
TreeNode: add missing setters

Even though they do nothing, it is best practice in Scout to use the
setters to update the state unless it is just a data object.

bschwarzent committed:
Bookmarks: static getter for BookmarkSupport

389569

bschwarzent committed:
Bookmarks: delegate display text to page

389569

bschwarzent committed:
Bookmarks: deserialize data objects from server

389569

cguglielmo committed:
Bookmarks: ensure every page has a uuid and pageParam

389577

bschwarzent committed:
Bookmarks: improvements for hybrid pages

Improve page loading without also selecting the page (e.g. during
bookmark activation):
- PageWithNodes: populate newly created detail table when children were
  loaded previously.
- PageWithTable: since the children are derived from the detail table,
  ensure that is automatically created when loadChildren() is called
  (otherwise, no data would be loaded).

BookmarkSupport:

Create search data, table preferences and chart filter config when
bookmarking JS and hybrid pages. For hybrid pages, a HybridAction is
used to compute the page state on the UI server. JS pages with a hybrid
search form are also supported.

Applying a bookmark is no longer asynchronous, since this caused issues
with code that did not wait for the promise to complete (e.g.
SuitePageExtension). If the newly applied page state should also cause
a reload of the data, use applyBookmarkToPageAndReload().

For hybrid pages it is now possible to apply the page state when opening
a bookmark from the UI server (-> applyParentBookmarkPage=true).

389569

bschwarzent committed:
Bookmarks: support update of js page definition without children loaded

389569

bschwarzent committed:
Bookmarks: extract bookmark creation to BookmarkDoBuilder

389569

cguglielmo committed:
Bookmarks: improve page resolving

PageResolver now caches the object types for pages with a dummy param
for a faster lookup.

Removed pageParamType because it is not needed so far.

389569

bschwarzent committed:
Bookmarks: refactor hybrid action

GetPageStateForBookmark -> CreateBookmark

389569

bschwarzent committed:
Bookmarks: cleanup FIXMEs

389569

cguglielmo committed:
ObjectFactory: add objectFactoryHints decorator

Until now, every object created using scout.create() got a unique id
if no id was passed by the model, unless ensureUniqueId was set to
false when calling scout.create().

Normally, the id is either passed by the ui server to the model
adapters. Or set manually by the developer in the model files so a
widget can be found using Widget.widget('id'). If a developer did not
assign a specific id for a Scout JS widget, a unique id was assigned
automatically. This is sometimes useful for example when using the
function arrays.union().

But some objects just never need an id. Whether an id should be assigned
automatically can now be controlled with the decorator
objectFactoryHints.
Currently, the decorator is present on Widget, TableRow, TreeNode and
ModelAdapter.

Similar to the id, sometimes it is not desired that the objectType
is written to the resulting object. This can now also be controlled
by the hints and is used by BaseDoEntity.

389577

bschwarzent committed:
Bookmarks: move bookmark-related data objects to scout

The API of BookmarkSupport was streamlined and documentation was added.
Bookmark creation and activation can now handle errors themselves or
delegate them to the caller.

When resolving child pages of table pages, in addition to the row
identifier the 'pageParam' can be used to identify matching pages. This
allows to omit the row identifier if it cannot be properly serialized,
e.g. for IDs.

BookmarkAdapter was removed as it was not really an adapter. The
remaining method were moved to BookmarkSupport instead.

389569

cguglielmo committed:
Session: enable inspector in dev mode

session.inspector should be true when running a Scout JS app in dev mode

For Scout Classic app this is done by the UiSession.java.

cguglielmo committed:
Uuid for Scout JS objects: simplify and improve

- rename uiId to uiSeqId to reduce confusion with uuid
- remove static methods if possible
  Since it is a singleton the methods should be on the instance for
  easier extensibility.
  uuidPathSkipWidgets needs to stay a static property because it needs
  to be accessed before the object factory is initialized.
- remove InspectorObjectIdProvider
  There is no need to send a separate classId for columns anymore.
  This would only be necessary if user preferences for Java tables were
  created by the UI. But since this is not the case, classId of columns
  don't need a special handling.
- ignore model adapter ids
  These ids are dynamic, so they must not be part of the uuid
- don't include outline as parent
  Because there is never more than one outline shown at a time,
  there is no need to include any outline in a uuidPath.
  Also, appending the outline even for pages which have a classId would
  break existing testing tools. It would be tolerable if it had a
  benefit, but there is none.
- add uuidPath option considerSkipWidgets
  If the starting element does not have a relevant id (uuid, classId or
  id), the elements cannot be skipped (no sparse path can be created)
  . For example: UI only elements like the ellipsis menu in a toolbar
  may appear multiple times on a form. If parents were skipped, the
  uuidPath would most likely be the same for those elements.
  To improve this, parents are now only skipped if the current element
  has a relevant id.
  Example UuidPath: EllipsisMenu|MenuBarBox|MenuBar|DetailBox|MyForm
  -> DetailBox has a relevant id, parents of the DetailBox may be skipped
  according to the skipWidgets list.
- remove uuidPath option appendParent
  It is not needed anymore
- don't include id and objectType
  It is not necessary to include both for the fallback uuid. If there
  is an id it should be unique in the context of the static model
  (e.g. form, template box, etc.). If there is no id, the object
  is most likely a top level element (form, template) with a custom
  objectType. Also, parent elements won't be skipped in that case, so
  the uuid path gets more precise. This makes the uuid paths more clear
  with less redundant information.
- don't consider top level ids anymore
  - Top level ids (e.g. of a form) were ignored so far and replaced by
    a uiSeqId. This means they were unique and could be used for object
    maps like the DesktopBench does. This would not work anymore.
  - Most forms should not have one because they were not of use.
  - Top level elements already have a unique objectType, the id would
    most likely consist of the same value
- enabled fallback handling for inspector
  Testing tools need a unique id to identify the widgets in the DOM
  and to record test cases. If the dom elements for the widgets
  automatically get a unique id without the need to assign one manually,
  the testability of Scout apps improves immediately. However, if the
  ids or objectTypes change, test cases may have to be adjusted. To
  minimize migration efforts, manually assigning uuids may still be
  a valid option.
- ui only elements now get a unique id
  Testing tools could not identify ui only elements like form field
  status, ellipsis menu, etc. Now, because the objectType is
  included, they get a unique id.
- add support for dependent uuids
  Some widgets like tabs or context-menu items are automatically created
  and not specified by the developer. Hence, they don't have an
  uuid or classId but still should be uniquely identifiable for testing
  tools. They now use the uuid or classId of their main element with a
  prefix. The prefix is necessary to avoid a conflict with the main
  element (e.g. to ensure the uuid is not added twice to the DOM).
- add more widgets to skipWidgets list
  Elements only used for layouting don't need to be included in the
  uuidPath, they don't make it better just longer.

389577

bschwarzent committed:
Bookmarks: add missing NLS texts

Also, cleanup some old Scout NLS texts.

389569

bschwarzent committed:
Bookmarks: various improvements

- Rename methods:
  "openBookmarkInOutline" -> "activateBookmark"
  "openBookmarkInPage"    -> "applyBookmarkToPage"
- Add resolveOutline() to BookmarkSupport
- Add generic timeout config property for UI callbacks
- Use $.when() to convert native promises to jquery promises
- Cleanup whitespace

389569

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+    return $.when(this._activateBookmark(bookmark, options));
+  }
+
+  async _activateBookmark(bookmark: IBookmarkDo, options?: ActivateBookmarkOptions): Promise<void> {
```
Missing protected modifier

cguglielmo commented on eclipse-scout-core/src/bookmark/ActivateRemainingBookmarkUiCallbackHandler.ts:
```diff
@@ -0,0 +1,31 @@
+
+export class ActivateRemainingBookmarkUiCallbackHandler implements UiCallbackHandler {
+
+  handle(param: UiCallbackParam): JQuery.Promise<void> {
```
Consider move this back because there is no caller in the Scout Java code

bschwarzent replied:
Scout implementation is now purely in JS. Hybrid support has to be added in a custom implementation.

cguglielmo commented on eclipse-scout-core/src/bookmark/ApplyBookmarkToJsPageUiCallbackHandler.ts:
```diff
@@ -0,0 +1,22 @@
+ */
+import {BookmarkSupport, Desktop, IBookmarkDo, Page, scout, UiCallbackHandler, UiCallbackParam} from '../index';
+
+export class ApplyBookmarkToJsPageUiCallbackHandler implements UiCallbackHandler {
```
Consider move this back, see other handler

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkForm.ts:
```diff
@@ -0,0 +1,33 @@
+import {BookmarkDo, BookmarkFormModel, BookmarkFormWidgetMap, Form, FormModel} from '../index';
+import model from './BookmarkFormModel';
+
+export class BookmarkForm extends Form implements BookmarkFormModel {
```
We need to discuss whether BookmarkForm and ManageBookmarksForm should be in Scout or JsWidgets only

bschwarzent replied:
Forms stay in Scout for now but use a customizable BookmarkStore to load/save data.

cguglielmo commented on eclipse-scout-core/src/bookmark/ManageBookmarksForm.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  protected override _save(data: any): JQuery.Promise<void> {
+    return bookmarks.storeAllBookmarks(this.bookmarks, 'jswidgets:bookmarks')
```
If this stays in Scout, it needs another id than jswidgets

bschwarzent replied:
Default ID is now 'scout:bookmarks:' + window.location.pathname (see LocalBookmarkStore.ts)

cguglielmo commented on eclipse-scout-core/src/bookmark/ManageBookmarksForm.ts:
```diff
@@ -0,0 +1,107 @@
+    this._editSelectedBookmark();
+  }
+
+  _editSelectedBookmark() {
```
Missing protected modifier

cguglielmo commented on eclipse-scout-core/src/bookmark/ManageBookmarksForm.ts:
```diff
@@ -0,0 +1,107 @@
+    form.whenSave().then(() => {
+      let bookmark = form.bookmark;
+      let name = bookmark.title;
+      // FIXME bsh [js-bookmark] merge with importData()
```
Solve FIXME

cguglielmo commented on eclipse-scout-core/src/bookmark/ManageBookmarksForm.ts:
```diff
@@ -0,0 +1,107 @@
+
+    this.widget('EditMenu').on('action', this._onEditMenuAction.bind(this));
+    this.widget('DeleteMenu').on('action', this._onDeleteMenuAction.bind(this));
+    // FIXME bsh [js-bookmark] Move MoveRowMenu to Scout
```
Solve FIXME

cguglielmo commented on eclipse-scout-core/src/bookmark/ManageBookmarksForm.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  protected override _load(): JQuery.Promise<any> {
+    return bookmarks.loadAllBookmarks('jswidgets:bookmarks')
```
If this stays in Scout, it needs another id

cguglielmo commented on eclipse-scout-core/src/bookmark/bookmarks.ts:
```diff
@@ -0,0 +1,114 @@
+
+export const bookmarks = {
+
+  // FIXME bsh [js-bookmark] Remove the following debugging/testing methods
```
Remove or cleanup to keep it for Scout JS archetype? I think it could be useful to have a default implementation that stores bookmarks to local storage

bschwarzent replied:
Utility was removed completely.

cguglielmo commented on eclipse-scout-core/src/bookmark/bookmarks.ts:
```diff
@@ -0,0 +1,114 @@
+    return dataObjects.parse(raw, Array<BookmarkDo>);
+  },
+
+  setBookmarkStore(bookmarkStore: BookmarkDo[], storeId?: string) {
```
Naming is confusing, maybe writeToStorage, readFromStorage?

cguglielmo commented on eclipse-scout-core/src/bookmark/bookmarks.ts:
```diff
@@ -0,0 +1,114 @@
+    });
+  },
+
+  // FIXME bsh [js-bookmark] Remove and replace with actual implementation
```
Remove fixme

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+    });
+    return builder.build()
+      .catch(error => {
+        this.handleCreateBookmarkError(error); // FIXME bsh [js-bookmark] Make this optional
```
Change FIXME to TODO if you don't plan to solve it for this release or even better, just remove it because it is not needed at the moment, or is it?

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+      });
+  }
+
+  createBookmarkForRefresh(options?: Omit<BookmarkDoBuilderModel, 'desktop'>): JQuery.Promise<IBookmarkDo> {
```
Please explain with JsDoc what for refresh means

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+    }
+  }
+
+  protected async _activateBookmarkHybrid(bookmark: IBookmarkDo, options?: ActivateBookmarkOptions): Promise<void> {
```
Would it be possible to remove the hybrid / classic code from Scout and only keep the Scout JS Code?

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+    let parentPage = param.parentPage;
+    let parentBookmarkPage = param.parentBookmarkPage;
+
+    // FIXME bsh [js-bookmark] HACKY-HACKY! Find a better solution.
```
Improve this FIXME comment, can we solve it at all? If no, remove it. If yes, explain what the problem is

bschwarzent replied:
Special handling is delegated to PageWithTable#setSearchFilter, which can then be customized by individual page implementations.

cguglielmo commented on eclipse-scout-core/src/bookmark/bookmarks.ts:
```diff
@@ -0,0 +1,114 @@
+ */
+import {arrays, BookmarkDo, BookmarkSupport, dataObjects, IBookmarkDo, UuidPool, webstorage} from '../index';
+
+export const bookmarks = {
```
Loading and storing bookmarks should be replacable -> create a class BookmarkStore and a LocalBookmarkStore

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+    helper.importConfig(page, bookmarkPage.chartTableControlConfig);
+  }
+
+  protected _prepareShowRelatedCustomerData(page: PageWithTable, bookmarkPage: TableBookmarkPageDo) {
```
This does not belong to Scout

bschwarzent replied:
Removed from scout

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+        searchForm.markAsSaved();
+      }
+    } else {
+      // FIXME bsh [js-bookmark] HACKY-HACKY! Replace by searchFilter property on page. How to instruct existing hybrid form to import this again?
```
remove HACKY-HACKY comment

cguglielmo commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+      });
+    }
+    let selectedKeys = selectedRows.map(row => row.getKeyValues());
+    // FIXME bsh [js-bookmark] Is this even required or should we just use selectRows()? It seems a bit awkward to convert the rows to keys and then back to rows.
```
May be necessary for hierarchical tables, restore selection expands the parents. Please verify and remove the fixme

bschwarzent replied:
Added Table#expandParentRows

cguglielmo commented on eclipse-scout-core/src/desktop/hybrid/hybridUtil.ts:
```diff
@@ -0,0 +1,113 @@
+ */
+import {HybridActionContextElement, HybridActionContextElementConverters, HybridActionContextElements, JsonHybridActionContextElement, ModelAdapter, scout, Session, Widget} from '../../index';
+
+export class HybridUtilImpl {
```
Remove Impl Suffix, Just name it HybridUtil

cguglielmo commented on eclipse-scout-core/src/desktop/outline/pages/Page.ts:
```diff
@@ -135,30 +180,46 @@ export class Page extends TreeNode implements PageModel {
     this.trigger('destroying');
     super._destroy();
-    if (this.detailTable) {
+    if (this.detailTable?.destroy) {
```
Explain why destroy may be missing: Probably because if minimalInit is called detailTable may still be the model instead of a real tabe

bschwarzent replied:
Is this even correct? Even when the "destroy" method is missing, shouldn't we set "this.detailTable" to null anyway? Maybe the `?.` should be moved to the next line instead.

cguglielmo commented on eclipse-scout-core/src/desktop/outline/pages/PageDetailMenuContributor.ts:
```diff
@@ -0,0 +1,66 @@
+    this.page = page;
+  }
+
+  abstract contribute(originalMenus: Menu[], detailContent: MenuOwner);
```
Add return type and js doc

cguglielmo commented on eclipse-scout-core/src/desktop/outline/pages/PageWithNodes.ts:
```diff
@@ -39,8 +39,16 @@ export class PageWithNodes extends Page {
   }
 
+  protected override _initDetailTable(table: Table) {
+    super._initDetailTable(table);
```
A test case for this scenario would be nice. Also fix typo apges

cguglielmo commented on eclipse-scout-core/src/desktop/outline/pages/ParentTablePageMenuContributor.ts:
```diff
@@ -0,0 +1,44 @@
+
+export class ParentTablePageMenuContributor extends PageDetailMenuContributor {
+
+  contribute(originalMenus: Menu[], detailContent: MenuOwner) {
```
Add return type

cguglielmo commented on eclipse-scout-core/src/desktop/outline/OutlineAdapter.ts:
```diff
@@ -251,72 +260,115 @@ export class OutlineAdapter extends TreeAdapter {
+      parent: pageModel.parent,
+      owner: pageModel.owner,
+      objectType: pageModel.jsPageObjectType,
+      // FIXME bsh [js-bookmark] Check which properties we want to explicitly inherit form the Java page
```
Remove this fixme, the properties should be fine

cguglielmo commented on eclipse-scout-core/src/desktop/Desktop.ts:
```diff
@@ -68,6 +68,7 @@ export class Desktop extends Widget implements DesktopModel, DisplayParent {
   url: URL;
   responsiveHandler: ResponsiveHandler;
   busySupport: BusySupport;
+  bookmarkSupport: BookmarkSupport;
```
Remove this as it is not used

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -4978,9 +4987,7 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
    */
   setLimitedResultTableStatus(limitedResult: boolean) {
     if (!limitedResult) {
-      if (this.tableStatus instanceof LimitedResultTableStatus) {
```
Ask mvi why this was done and/or test it. Is it possible that an existing tableStatus is now removed? tableStatus may be set by ui server

cguglielmo replied:
If there was an error status because of a loading error, it did not vanish anymore even if data could be loaded later. Custom error status would be replaced, yes, but the same happens in Java -> not supported yet

cguglielmo commented on eclipse-scout-core/src/testing/menu/MenuSpecHelper.ts:
```diff
@@ -16,11 +16,10 @@ export class MenuSpecHelper {
   }
 
-  createModel(text?: string, iconId?: string, menuTypes?: string[]): FullModelOf<Menu> {
+  createModel(text?: string, menuTypes?: string[]): FullModelOf<Menu> {
```
Adjust commit comment and add that MenuSpecHelper was simplified by removing iconId

cguglielmo commented on eclipse-scout-core/src/scout.ts:
```diff
@@ -41,6 +45,11 @@ export type ObjectOrModel<T> = T | ModelOf<T>;
+/**
+ * Represents an instance of an object or its type ({@link ObjectType}).
+ */
+export type ObjectOrType<T> = T | ObjectType<T>;
```
Adjust commit comment and add that ObjectOrType was added

cguglielmo commented on org.eclipse.scout.rt.ui.html/src/main/java/org/eclipse/scout/rt/ui/html/json/table/JsonTable.java:
```diff
@@ -574,6 +575,7 @@ public JSONObject toJson() {
       json.put(PROP_FILTERS, filtersToJson(getModel().getUserFilterManager().getFilters()));
     }
     json.put(PROP_HAS_RELOAD_HANDLER, getModel().getReloadHandler() != null);
+    json.put(PROP_USER_PREFERENCE_CONTEXT, getModel().getUserPreferenceContext());
```
Why is this necessary? It is not used in JS

bschwarzent commented on org.eclipse.scout.rt.app/src/main/java/org/eclipse/scout/rt/app/Application.java:
```diff
@@ -169,6 +169,12 @@ protected void logServerReady() {
       sb.append("  ").append(protocol).append("://").append(ip).append(":").append(port).append(contextPath).append('\n');
     }
     sb.append("---------------------------------------------------------------------\n");
+    try {
```
TODO BSH: extract to separate commit

bschwarzent commented on org.eclipse.scout.rt.ui.html/src/main/java/org/eclipse/scout/rt/ui/html/json/InspectorInfo.java:
```diff
@@ -9,24 +9,100 @@
+  /**
+   * Regex for a UUID as created by {@link UUID}.
+   */
+  private static final String UUID_PATTERN = "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}";
```
private -> protected

bschwarzent commented on org.eclipse.scout.rt.ui.html/src/main/java/org/eclipse/scout/rt/ui/html/json/InspectorInfo.java:
```diff
@@ -9,24 +9,100 @@
+      return MessageDigest.getInstance("SHA-256");
+    }
+    catch (NoSuchAlgorithmException e) {
+      throw new ProcessingException("Unable to create SHA-256 message digest.", e);
```
This may throw an exception during loading of the class. Is this good? How expensive is MessageDigest.getInstance() really? Maybe we can just inline it where we need it, or convert this function to a non-static method.

bschwarzent commented on org.eclipse.scout.rt.ui.html/src/main/java/org/eclipse/scout/rt/ui/html/json/InspectorInfo.java:
```diff
@@ -9,24 +9,100 @@
+   *     The classId to prepare. Must not be {@code null}.
+   * @return The id ready to be sent to the browser.
+   */
+  public String prepareClassId(String classId) {
```
Is it useful to have this as a public method?

bschwarzent commented on org.eclipse.scout.rt.ui.html/src/main/java/org/eclipse/scout/rt/ui/html/json/InspectorInfo.java:
```diff
@@ -9,24 +9,100 @@
+   * @param model
+   *     The model for which the properties should be added. May be {@code null} then this method does nothing.
+   */
+  public <T> void put(HttpServletRequest req, JSONObject json, T model) {
```
Instead of passing the http request, consider a boolean flag instead, e.g. `addModelClass`. The API feels strange right now, since we are not "putting" something on the http request, but on the json object. The http qurest is only used to determine the inspector hint.

bschwarzent commented on org.eclipse.scout.rt.ui.html/src/main/java/org/eclipse/scout/rt/ui/html/json/desktop/JsonOutline.java:
```diff
@@ -53,6 +56,9 @@ public class JsonOutline<OUTLINE extends IOutline> extends JsonTree<OUTLINE> {
   private static final String PROP_SHOW_TILE_OVERVIEW = "showTileOverview";
   private static final String PROP_COMPACT_ROOT = "compactRoot";
 
+  public static final String EVENT_SEARCH_FILTER_FOR_PAGE_REQUEST = "searchFilterForPageRequest";
```
TODO BSH: These events (and everthing connected to them) seem to be obsolete. Check and remove!

bschwarzent commented on eclipse-scout-core/src/dataobject/StringValueDo.ts:
```diff
@@ -0,0 +1,15 @@
+import {BaseDoEntity, typeName, ValueDo} from '..';
+
+@typeName('scout.StringValue')
+export class StringValueDo extends BaseDoEntity implements ValueDo<string> {
```
Why does this exist in JS, but all other Java subclasses of IValueDo do not?

bschwarzent commented on eclipse-scout-core/src/desktop/outline/pages/PageModel.ts:
```diff
@@ -1,22 +1,30 @@
-export interface PageModel extends TreeNodeModel {
+export interface PageModel extends TreeNodeModel, ObjectModelWithUuid<TreeNode> {
+  /**
+   * The {@link PageParamDo} containing all parameters that are required to create the page.
```
```suggestion
   * The {@link PageParamDo} containing all parameters that are required to load the page.
```
(Creating the page instance requires more data, e.g. the parent)

bschwarzent commented on eclipse-scout-core/src/bookmark/BookmarkSupport.ts:
```diff
@@ -0,0 +1,544 @@
+   * If the app does not have any active sessions (e.g. during unit testing), this argument is mandatory.
+   */
+  static get(session?: Session): BookmarkSupport {
+    session = session || App.get().sessions[0];
```
TODO BSH: Simplify this, see PageResolver.get()

bschwarzent commented on eclipse-scout-core/src/desktop/outline/pages/ParentTablePageMenuContributor.ts:
```diff
@@ -0,0 +1,44 @@
+    return [...parentTablePageMenus, ...originalMenus];
+  }
+
+  protected _computeParentTablePageMenus(newParent: Widget): Menu[] {
```
Why is the parameter called "newParent"?

bschwarzent commented on eclipse-scout-core/src/desktop/outline/pages/ParentTablePageMenuContributor.ts:
```diff
@@ -0,0 +1,44 @@
+      .map(menu => this._cloneMenu(menu, newParent));
+  }
+
+  protected _filterParentTablePageMenus(tablePageMenus: Menu[]): Menu[] {
```
This name does not really explain what the method does. Maybe rename to "_getParentTablePageMenusForClone"?

bschwarzent commented on eclipse-scout-core/src/desktop/outline/OutlineAdapter.ts:
```diff
@@ -251,72 +260,115 @@ export class OutlineAdapter extends TreeAdapter {
+        return this.modelAdapter._createJsPage(pageModel, this._createTreeNodeOrig.bind(this));
+      } catch (error) {
+        // Create a broken page instead of showing a fatal error so the application can still be used.
+        pageModel = {
```
Maybe extract the creation of the error page to a method?

bschwarzent commented on eclipse-scout-core/src/desktop/outline/OutlineAdapter.ts:
```diff
@@ -251,72 +260,115 @@ export class OutlineAdapter extends TreeAdapter {
+    pageModel = this.modelAdapter._initNodeModel(pageModel);
+    if (pageModel.nodeType === 'jsPage') {
+      try {
+        return this.modelAdapter._createJsPage(pageModel, this._createTreeNodeOrig.bind(this));
```
Why is the real page returned as an instance, but the error page only as a model? Can we also create an instance for the error page?

cguglielmo replied:
Because createJsPage does an assertion for classId at the end for which it needs to be a real page. The broken page will be created "automatically" at the end of the createTreeNodeRemote function, so it is not necessary

bschwarzent commented on eclipse-scout-core/src/desktop/outline/OutlineAdapter.ts:
```diff
@@ -251,72 +260,115 @@ export class OutlineAdapter extends TreeAdapter {
+    return page;
+  }
+
+  protected _initJsPageModel(pageModel: PageModel): PageModel {
```
Is this method needed? Could be inlined into _createJsPage().

bschwarzent commented on eclipse-scout-core/src/desktop/outline/OutlineAdapter.ts:
```diff
@@ -251,72 +260,115 @@ export class OutlineAdapter extends TreeAdapter {
+  }
+
+  protected _initJsPageModel(pageModel: PageModel): PageModel {
+    let pageParam = dataObjects.deserialize(pageModel.pageParam);
```
Is PageParam deserialized 2x? Here and in _initNodeModel(). Consider remove this in favor of _initNodeModel.

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+/**
+ * Helper class to extract IDs of objects and to compute uuidPaths.
+ */
+export class ObjectUuidProvider implements ObjectUuidProviderModel, ObjectWithType {
```
Since it does not only provide UUIDs, maybe rename to ObjectIdProvider?

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+  /**
+   * Prefix for all UI generated IDs.
+   */
+  static UI_SEQ_ID_PREFIX = '_ui_'; // must not contain any dots ('.') so that the id can be used as css selector "#..." and for the RegExp 'UI_SEQ_PATTERN'.
```
This prefix looks a bit ugly. Maybe simplify?

- `_ui_124`
- `ui-1234`
- `ui_1234`
- `_ui1234`
- `_ui-1234`
- `ui__1234`

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+  objectType: string;
+  id: string;
+
+  protected _uiSeqIdNo = 0;
```
Maybe remove the "No" suffix? Or keep it but remove the "Id" bit instead? Currently, the same seems a bit weird.

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+   */
+  static DEPENDENT_UUID_DELIMITER = '@';
+
+  /** use {@link createUiSeqId} to generate a new ID */
```
What is the use of this jsdoc? How is it linked to the regex?

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+  static uuidPathAlwaysSkipRules: ((widget: Widget) => boolean)[] = [];
+
+  constructor() {
+    this.objectType = null;
```
Why do we need these two properties? The Object is a singleton.

cguglielmo replied:
objectType is added by the object factory. Id should be removed

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+   * @returns the uuid path starting with this object's uuid or null if no path can be created.
+   */
+  uuidPath(object: ObjectUuidSource, options?: UuidPathOptions) {
+    options = scout.nvl(options, {});
```
Maybe specify the default value in the method signature instead.

cguglielmo replied:
scout.nvl also considers null in addition to undefined. So uuidPath(object, null) would cause an error. This is why I don't use it very often. We should discuss it

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+    // Prepare options for the next iteration
+    options.considerSkipWidgets = considerSkipWidgets;
+    options.abortIfNoUuidFound = scout.nvl(options.abortIfNoUuidFound, false); // Skip parents without an uuid
+    options.parent = null; // Passed element must only be used for starting element
```
Not sure what this comment means. Maybe rephrase?

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+  protected _computeConsiderSkipWidgets(options: UuidPathOptions, object: ObjectUuidSource): UuidPathConsiderSkipWidgets {
+    if (objects.isNullOrUndefined(options.considerSkipWidgets) || options.considerSkipWidgets === 'dynamicFalse') {
+      // Do not skip parent widgets if the object only has an object type because the objectType normally is not unique enough
+      return (!!object.uuid || !!object.classId || this._considerId(object)) ? 'dynamicTrue' : 'dynamicFalse';
```
`!!` not necessary here, since the values are not returned directly.

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+   * @param useFallback Optional boolean specifying if a fallback identifier may be created in case an object has no specific identifier set. The fallback may be less stable. Default is true.
+   * @returns the uuid for the object or null.
+   */
+  uuid(object: ObjectUuidSource, useFallback?: boolean): string {
```
Specify default value of `useFallback` in method signature.

bschwarzent commented on eclipse-scout-core/src/util/ObjectUuidProvider.ts:
```diff
@@ -0,0 +1,326 @@
+    return strings.join(ObjectUuidProvider.DEPENDENT_UUID_DELIMITER, prefix, uuid);
+  }
+
+  setDependentUuid(prefix: string, source: ObjectUuidSource, target: Required<ObjectUuidSource>): string {
```
Add jsdoc

bschwarzent commented on eclipse-scout-core/src/form/fields/groupbox/GroupBox.ts:
```diff
@@ -865,3 +865,5 @@ export class GroupBox extends CompositeField implements GroupBoxModel {
 export type GroupBoxBorderDecoration = EnumObject<typeof GroupBox.BorderDecoration>;
 export type GroupBoxMenuBarPosition = EnumObject<typeof GroupBox.MenuBarPosition>;
+
+ObjectUuidProvider.uuidPathSkipWidgets.add(GroupBox);
```
Could we specify this on CompositeField instead of individually on GroupBox, TabBox etc?

cguglielmo replied:
No, see jsdoc of ObjectUuidProvider.uuidPathSkipWidgets

bschwarzent commented on eclipse-scout-core/src/form/fields/wizard/WizardProgressField.ts:
```diff
@@ -112,8 +114,8 @@ export class WizardProgressField extends FormField implements WizardProgressFiel
+      inspector.applyInfo(step, $step, this.session);
       if (this.session.inspector) {
-        inspector.applyInfo(step, $step);
         $step.attr('data-step-index', step.index);
```
This could probably be safely added always without checking for `this.session.inspector`.

bschwarzent commented on eclipse-scout-core/src/tabbox/SimpleTab.ts:
```diff
@@ -85,6 +85,9 @@ export class SimpleTab<TView extends SimpleTabView = SimpleTabView> extends Widg
     this.saveNeededVisible = (this.view ? this.view.saveNeededVisible : model.saveNeededVisible);
     this.status = (this.view ? this.view.status : model.status);
+    this.uuid = model.uuid;
+    this.classId = model.classId;
```
Why is this the only widget that does this manually? Isn't this already done by the super call?

bschwarzent commented on eclipse-scout-core/src/testing/JasmineScoutUtil.ts:
```diff
@@ -130,5 +130,90 @@ export const JasmineScoutUtil = {
+    options = options || {};
+    let pagesNotRequiringUuid: Set<Constructor<Page> | AbstractConstructor<Page>> = new Set([PageWithNodes, PageWithTable, AutoLeafPageWithNodes, ...options.pagesNotRequiringUuid || []]);
+    let pagesNotRequiringPageParam: Set<Constructor<Page> | AbstractConstructor<Page>> = new Set([PageWithNodes, PageWithTable, AutoLeafPageWithNodes, ...options.pagesNotRequiringPageParam || []]);
+    let missingUuids = new Set();
```
Specify type for these sets as well.

bschwarzent commented on eclipse-scout-core/src/testing/JasmineScoutUtil.ts:
```diff
@@ -130,5 +130,90 @@ export const JasmineScoutUtil = {
+   */
+  assertPageCompleteness(options?: PageCompletenessOptions) {
+    options = options || {};
+    let pagesNotRequiringUuid: Set<Constructor<Page> | AbstractConstructor<Page>> = new Set([PageWithNodes, PageWithTable, AutoLeafPageWithNodes, ...options.pagesNotRequiringUuid || []]);
```
The pattern `Constructor<Page> | AbstractConstructor<Page>` is a bit ugly. Can we simplfy this? Maybe only specify AbstractConstructor, it seems to cover "normal" constructors as well. Or rename `AbstractConstructor` to `Class`?

cguglielmo replied:
As discussed: It is ugly, but we postpone this topic. We would need 3 types and decide for every usage which is the correct one.

bschwarzent commented on eclipse-scout-core/src/testing/JasmineScoutUtil.ts:
```diff
@@ -130,5 +130,90 @@ export const JasmineScoutUtil = {
+
+    if (missingUuids.size > 0) {
+      fail(`Found page(s) without a uuid. Please ensure every page has a uuid.
+      ${Array.from(missingUuids).join('\n')}`);
```
The second line of this string has leading spaces, is this intended?

The syntax for multiline strings is a bit ugly. Do we really need them? I think classic strings would work better here. Or something like this:

```
fail([
  `Found ${missingUuids.length} pages without a uuid. Please ensure every page has a uuid.`,
  ...missingUuids
].join(\n'));
```

bschwarzent commented on eclipse-scout-core/src/util/inspector.ts:
```diff
@@ -1,28 +1,52 @@
   }
 };
+
+export interface InspectorModel extends Partial<ObjectWithUuid> {
```
Is the name "Model" appropriate here? Maybe "Options" or "Param" would be better?

bschwarzent commented on eclipse-scout-core/src/util/inspector.ts:
```diff
@@ -1,28 +1,52 @@
+   * element is undefined or the inspector is disabled in the session.
    */
-  applyInfo(model: { $container?: JQuery; id?: string; modelClass?: string; classId?: string }, $container?: JQuery) {
+  applyInfo(model: InspectorModel, $container?: JQuery, session?: Session) {
```
Are the arguments $container and session really needed when they can already be part of the first argument?

cguglielmo replied:
This is nothing new, it simplifies the usage

bschwarzent commented on eclipse-scout-core/src/util/inspector.ts:
```diff
@@ -1,28 +1,52 @@
+      return;
+    }
+
+    let modelClass = model.modelClass || (typeof model.objectType === 'string' && model.objectType);
```
`(typeof model.objectType === 'string' && model.objectType)` --> Is the second condition necessary?

bschwarzent commented on eclipse-scout-core/src/ObjectFactory.ts:
```diff
@@ -130,23 +130,31 @@ export class ObjectFactory {
+   * If not defined otherwise by the hints, the objectType is written to the created object.
+   *
+   * @param objectTypeOrModel This argument can be
+   *        <ul>
```
Use JsDoc notation (`-` instead of `<li>`)

bschwarzent committed:
Add AnyJsonValue marker type

When creating DOs with scout.create(), we get a TS compile error when
an attribute is typed as JsonValue:

  TS2589: Type instantiation is excessively deep and possibly infinite.

As workaround, we add a new type AnyJsonValue, which can be used in
data object classes instead. It is equivalent to "any", which is a bit
too permissive, but at least we can _semantically_ express that only
data that is valid in JSON should be assigned to these attributes.

389577

bschwarzent committed:
Bookmarks: add bookmark store

- Add BookmarkStore API to load/store bookmarks.
- A default implementation that reads/writes from the browser's local
  storage is provided (LocalBookmarkStore).
- Improve default bookmark management forms.

389569
